daemon/src/engine/client/cl_parse.cpp

```diff
@@ -108,148 +84,83 @@ CL_ParsePacketEntities
-	oldstate = nullptr;
-
-	if ( !oldframe )
-	{
-		static clSnapshot_t nullframe{};
-        nullframe.valid = false;
-
-		oldframe = &nullframe;
-		oldnum = MAX_GENTITIES; // guaranteed out of range
-	}
-	else
-	{
-		if ( oldindex >= oldframe->numEntities )
-		{
-			oldnum = MAX_GENTITIES;
-		}
-		else
-		{
-			oldstate = &cl.parseEntities[
-			             ( oldframe->parseEntitiesNum + oldindex ) & ( MAX_PARSE_ENTITIES - 1 ) ];
-			oldnum = oldstate->number;
-		}
-	}
-
-	while ( 1 )
-	{
-		// read the entity index number
-		newnum = MSG_ReadBits( msg, GENTITYNUM_BITS );
-
-		if ( newnum >= ( MAX_GENTITIES - 1 ) )
-		{
-			break;
-		}
-
-		if ( msg->readcount > msg->cursize )
-		{
-			Com_Error( ERR_DROP, "CL_ParsePacketEntities: end of message" );
-		}
-
-		while ( oldnum < newnum )
-		{
-			// one or more entities from the old packet are unchanged
-			if ( cl_shownet->integer == 3 )
-			{
-				Com_Printf( "%3i:  unchanged: %i\n", msg->readcount, oldnum );
-			}
-
-			CL_DeltaEntity( msg, newframe, oldnum, oldstate, true );
-
-			oldindex++;
-
-			if ( oldindex >= oldframe->numEntities )
-			{
-				oldnum = MAX_GENTITIES;
-			}
-			else
-			{
-				oldstate = &cl.parseEntities[
-				             ( oldframe->parseEntitiesNum + oldindex ) & ( MAX_PARSE_ENTITIES - 1 ) ];
-				oldnum = oldstate->number;
-			}
-		}
-
-		if ( oldnum == newnum )
-		{
-			// delta from previous state
-			if ( cl_shownet->integer == 3 )
-			{
-				Com_Printf( "%3i:  delta: %i\n", msg->readcount, newnum );
-			}
-
-			CL_DeltaEntity( msg, newframe, newnum, oldstate, false );
-
-			oldindex++;
-
-			if ( oldindex >= oldframe->numEntities )
-			{
-				oldnum = MAX_GENTITIES;
-			}
-			else
-			{
-				oldstate = &cl.parseEntities[
-				             ( oldframe->parseEntitiesNum + oldindex ) & ( MAX_PARSE_ENTITIES - 1 ) ];
-				oldnum = oldstate->number;
-			}
+    // The entity packet contains the delta between the two snapshots, with data only
+    // for entities that were created, changed or removed. Entities entries are in
+    // order of increasing entity number, as are entities in a snapshot. Using this we
+    // have an efficient algorithm to create the new snapshot, that goes over the old
+    // snapshot once from the beginning to the end.
+
+    // If we don't have an old snapshot or it is empty, we'll recreate all entities
+    // from the baseline entities as setting oldEntityNum to MAX_GENTITIES will force
+    // us to only do step (3) below.
+    unsigned int oldEntityNum = MAX_GENTITIES;
+    if (oldSnapshot && oldSnapshot->entities.size() > 0){
+        oldEntityNum = oldSnapshot->entities[0].number;
+    }
+
+    // Likewise when we don't have an old snapshot, oldEntities just has to be an empty
+    // vector so that we skip step (4)
+    std::vector<entityState_t> dummyEntities;
+    auto& oldEntities = oldSnapshot? oldSnapshot->entities : dummyEntities;
+    auto& newEntities = newSnapshot->entities;
+
+    unsigned int numEntities = MSG_ReadShort(msg);
```

Comment:
MSG_ReadShort returns int, is the cast safe? can it ever return negative values?